keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+#### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: JobRecreatePodsWhenFailed
+  - Components depending on the feature gate: kube-controller-manager
+
+###### Does enabling the feature change any default behavior?
+
+For enabling JobRecreatePodsWhenFailed:
+
+a) Count the number of terminating pods and populate in JobStatus
+b) With `RecreationPodsWhen: Failed` specified, pods will only be recreated when they are fully terminated.
+
+This could potentially make jobs (where pods are terminated) slower due to waiting for terminating pods
+to be fully deleted.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes.
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Terminating pods will no longer be tracked so terminating pods will be considered deleted and new pods will be created.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes. Unit tests will include the fields off/on and verify behavior.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+#### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+#### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
```

Comment:
This section is a must.

Reply:
@alculquicondor or @mimowo any suggestions here?  Not exactly sure what metrics we could use or what exists? 

Is sync_duration one that could be used here?  I would expect the sync_duration to take longer if the feature toggle is on.

Reply:
yeah, in general sync_duration could give some signal about performance regressions, but I think more important is `job_syncs_total`, as it might show an increased number of failures when the feature is enabled.

Those are the generic metrics to use.

More specific to this feature could be `job_pods_finished_total`, compared to the metric for pod creation (is there one? I don't think this is incrementing any metric)

Next, you have to think whether new metrics make sense, but you have to think in an aggregated manner (all jobs, as opposed to per job). Other than pod creation, I don't think we need a new one.

Reply:
With what @mimowo has suggested I wonder if a avg metric between terminating and creation may make sense?  

I added a note to mention what metrics could signal an issue.